### PR TITLE
Add `floyd_warshall`

### PR DESCRIPTION
**Pull request by eriknw**
CC @jim22k @SultanOrazbayev @LuisFelipeRamos

I made a few minor modifications to the algorithms from what we wrote together today. I'm happy to answer any questions. I tinkered around a little to make things faster.

We can probably get this to work with `dask-graphblas` too, which I think would be pretty interesting, because it can create a massive, distributed matrix. It may not be the _best_ way to compute APSP, but _a_ way is better than no way at all :)

See the original LAGraph version of Floyd-Warshall here:
https://github.com/GraphBLAS/LAGraph/blob/ed55a49ee7138d2b5a6c5eb4329ccd0bf9e4ac17/old/experimental_algorithm/LAGraph_FW.c

I'll try to benchmark and compare this with a NumPy implementation on a beefy machine with lots of memory.

### CODE REVIEW

**Comment by SultanOrazbayev** [comment]
A potential optimization is to loop over n, where n is rows with non-zero values.

**Comment by eriknw** [comment]
> A potential optimization is to loop over n, where n is rows with non-zero values.

Good suggestion. I added an optimization where we only iterate over vertices that have nonempty rows _and_ nonempty columns. I _think_ this behaves correctly, but would appreciate if somebody could verify it.

I also introduced another temporary matrix to hold the outer product. We then drop the diagonal values from them.

All this performs better or similar in my limited benchmarking. Our strategy of keeping things sparse is _probably_ reasonable, because what if there are multiple groups of connected components? The final result frequently may not be sparse.

With the goal of "keeping things sparse", I wonder if there are any heuristics we could employ, such as iterating over vertices with small degrees first.

File: graphblas_algorithms/algorithms/shortest_paths/dense.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    # By using `offdiag` instead of `G._A`, we ensure that D will not become dense.
+    # Dense D may be better at times, but not including the diagonal will result in less work.
+    # Typically, Floyd-Warshall algorithms sets the diagonal of D to 0 at the beginning.
+    # This is unnecessary with sparse matrices, and we set the diagonal to 0 at the end.
+    # We also don't iterate over index `i` if either row i or column i are empty.
+    if G.is_directed():
+        A, row_degrees, column_degrees = G.get_properties("offdiag row_degrees- column_degrees-")
+        nonempty_nodes = binary.pair(row_degrees & column_degrees).new(name="nonempty_nodes")
+    else:
+        A, nonempty_nodes = G.get_properties("offdiag degrees-")
```
Comment (eriknw) [inline graphblas_algorithms/algorithms/shortest_paths/dense.py:18]: Note that we use some shorthand notation here. `"degrees-"` _does not_ include self-edges (i.e., diagonals), but `"degrees+"` _does_ include self-edges.

File: graphblas_algorithms/algorithms/shortest_paths/dense.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    if is_weighted:
+        D << A
+    else:
+        D(A.S) << 1  # Like `D << unary.one[int](A)`
+    del A
+
+    Row = Matrix(dtype, nrows=1, ncols=n, name="Row")
+    Col = Matrix(dtype, nrows=n, ncols=1, name="Col")
+    Outer = Matrix(dtype, nrows=n, ncols=n, name="temp")
+    for i in nonempty_nodes:
```
Comment (SultanOrazbayev) [inline graphblas_algorithms/algorithms/shortest_paths/dense.py:35]: Looks good!

**Comment by SultanOrazbayev** [comment]
> With the goal of "keeping things sparse", I wonder if there are any heuristics we could employ, such as iterating over vertices with small degrees first.

Not sure about this, it could lead to gotchas if users don't expect sorting to happen. (and sorting itself could be an issue, e.g. when using distributed graphs)

**Pull request merged**